src/utils/uintarray-encoding.js

```diff
@@ -5,30 +5,168 @@
+    result += leadingZeroDigit;
+    result += encodeUint(numbers[i]);
+  }
+  return result;
+}
+
+export function decodeUintArrayFromUrlComponent(s: string): number[] {
+  const array = [];
+  let i = 0;
+  while (i < s.length) {
+    const { value, hasLeadingZero, nextI } = decodeUint(s, i);
+    if (hasLeadingZero && array.length >= 1) {
+      const startValue = array[array.length - 1];
+      const endValue = value;
+      if (endValue > startValue) {
+        for (let x = startValue + 1; x < endValue; x++) {
+          array.push(x);
+        }
+      } else {
+        for (let x = startValue - 1; x > endValue; x--) {
+          array.push(x);
+        }
+      }
+    }
+    array.push(value);
+    i = nextI;
+  }
+  return array;
+}
+
+// Returns the number of items at numbers[start..] that can be substituted with a
+// "consecutive range" marker. If the return value is non-zero, there is a
+// consecutive range which starts at numbers[start - 1] and ends at
+// numbers[start + returnValue + 1].
+// A consecutive range is a range of numbers which either all go up by one or down by one.
+//
+// Example:
+//   We want to turn the sequence "1, 3, 4, 5, 6, 5, 4, 3, 2" into the "collapsed"
+//   sequence "1, 3, ...6, ...2".
+//   We skip "4, 5" and substitute them with a range marker:
+//   countSkippableConsecutiveNumbersAt([1, 3, 4, 5, 6, 5, 4, 3, 2], 2) === 2
+//                                             ^^^^ can be skipped
+//   We also skip "5, 4, 3":
+//   countSkippableConsecutiveNumbersAt([1, 3, 4, 5, 6, 5, 4, 3, 2], 5) === 3
+//                                                      ^^^^^^^ can be skipped
+function countSkippableConsecutiveNumbersAt(
+  numbers: number[],
+  start: number
+): number {
+  if (start < 1 || start + 1 >= numbers.length) {
+    return 0;
+  }
+  const previous = numbers[start - 1];
+  const current = numbers[start];
+  const next = numbers[start + 1];
+
+  let skipCount = 0;
+  if (current === previous + 1 && next === current + 1) {
+    // Found increasing consecutive range.
+    skipCount = 1;
+    while (
+      start + skipCount + 1 < numbers.length &&
+      numbers[start + skipCount + 1] === current + skipCount + 1
+    ) {
+      skipCount++;
+    }
+  } else if (current === previous - 1 && next === current - 1) {
+    // Found decreasing consecutive range.
+    skipCount = 1;
+    while (
+      start + skipCount + 1 < numbers.length &&
+      numbers[start + skipCount + 1] === current - skipCount - 1
+    ) {
+      skipCount++;
+    }
+  }
+  return skipCount;
+}
+
+const encodingDigits: string =
   '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ._';
 
-function uintToString(value: number): string {
+// Prepending this digit to an encoded number does not change the number: It's
+// like a leading zero. It has the continuation bit set to 1 but the value bits
+// set to zero.
+const leadingZeroDigit: string = encodingDigits[0b100000];
```

Comment:
I would move these 2 consts before all functions, because we start seeing them in some functions above. Also what do you think of making them all_caps like we do with constants usually?

Reply:
ok, sure